homeassistant/components/hassmpris/media_player.py

```diff
@@ -0,0 +1,625 @@
+            raise HomeAssistantError("cannot next: %s" % exc) from exc
+
+    async def async_media_previous_track(self) -> None:
+        """Skip to previous track."""
+        try:
+            await self.coordinator.client.previous(self.coordinator_context)
+        except Exception as exc:
+            raise HomeAssistantError("cannot previous: %s" % exc) from exc
+
+    async def async_media_seek(self, position: float) -> None:
+        """Send seek command."""
+        try:
+            trackid = self._metadata.get("mpris:trackid", None)
+            await self.coordinator.client.set_position(
+                self.coordinator_context,
+                trackid,
+                position,
+            )
+        except Exception as exc:
+            raise HomeAssistantError("cannot seek: %s" % exc) from exc
+
+    @property
+    def extra_state_attributes(self) -> dict[str, Any]:
+        """Return entity specific state attributes."""
+        if self.state == MediaPlayerState.OFF:
+            return {}
+        return {ATTR_PLAYBACK_RATE: self._attr_playback_rate}
+
+
+class MPRISCoordinator(
+    DataUpdateCoordinator[
+        dict[str, list[Union[mpris_pb2.MPRISPlayerUpdate, UNAVAILABLE]]]  # type: ignore[valid-type]
+    ]
+):
+    """The entity manager manages MPRIS media player entities.
+
+    This class is responsible for maintaining the known player entities
+    in sync with the state as reported by the server, as well as keeping
+    tabs of newly-appeared players and players that have gone.
+    """
+
+    config_entry: ConfigEntry
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        config_entry: ConfigEntry,
+        mpris_client: hassmpris_client.AsyncMPRISClient,
+        async_add_entities: AddEntitiesCallback,
+    ) -> None:
+        """Initialize the entity manager.
+
+        Arguments:
+          hass: the HomeAssistant singleton
+          config_entry: the configuration entry associated with
+                        this component (or integration?)
+          mpris_client: the MPRIS client endpoint object
+          async_add_entities: callback to add entities async
+        """
+        super().__init__(hass, _LOGGER, name="MPRIS")
+        self.data: dict[str, list[Union[mpris_pb2.MPRISPlayerUpdate, UNAVAILABLE]]] = {}  # type: ignore[valid-type]
+        self.config_entry = config_entry
+        self.async_add_entities = async_add_entities
+        self._client = mpris_client
+        self._shutdown: asyncio.Future[bool] = asyncio.Future()
+        self._started = False
+        self._deferred_task: asyncio.Task | None = None
+
+    @property
+    def client(self) -> hassmpris_client.AsyncMPRISClient:
+        """Return the MPRIS client associated with this entity manager."""
+        return self._client
+
+    async def start(self) -> None:
+        """Start the entity manager as a separate task."""
+        self.hass.loop.create_task(self.run())
+
+    async def run(self) -> None:
```

Comment:
This looks as if this code is a port of a thread-based implementation, is that the case? If it is, why is this not handled by the normal coordinator update code?

Reply:
It was never thread-based.

Q: what "normal coordinator update code" are you referring to?

The coordinator code *I* have read in `DataUpdateCoordinator` is structured as a poll-based mechanism.  I can't use that, because this is `async` streaming from the server as the server sends updates in real-time.  Best we can do is call `async_set_updated_data(...)` when it's time to send data to entities, which we do -- in a coalescer function called `_dispatch_deferred_update()`.

Reply:
Let's continue this conversation in the other thread about the `DataUpdateCoordinator` down below.